Excel export customization updates - docs plus kb for using RadSpreadProcessing (#899)
telerik/blazor-docs: knowledge-base/grid-custom-cell-formatting-with-radspreadprocessing.md

---
title: Custom cell formatting of the exported file with RadSpreadProcessing
description: How to format the cells of the exported file with RadSpreadProcessing?
type: how-to

> page_title: Custom cell formatting of the exported file with RadSpreadProcessing

Reviewer: ```suggestion
page_title: Custom cell formatting of exported Excel files with RadSpreadProcessing
```

slug: grid-kb-custom-cell-formatting-with-radspreadprocessing
position: 
tags: grid, custom, cell, format, radspreadprocessing
ticketid: 1559794
res_type: kb
---

## Environment

> <table>
> 	<tbody>
> 		<tr>
> 			<td>Product</td>
> 			<td>Grid for Blazor</td>
> 		</tr>
> 	</tbody>
> </table>

Reviewer: ```suggestion
<table>
    <tbody>
        <tr>
            <td>Product</td>
            <td>Grid for Blazor</td>
        </tr>
    </tbody>
</table>
```

## Description

> I want to format the cells of the exported file. For example, set background color. How to achieve this?

Reviewer: ```suggestion
I want to format the cells of the exported Excel file. For example, set background color. How to achieve this?
```

## Solution

> Handle the [OnAfterExport]({%slug grid-export-events%}#onafterexport) event of the Grid. It fires before the actual file is provided to the user. The `Stream` field of its event arguments contains the output of the export as `MemoryStream`.

Reviewer: ```suggestion
Handle the [OnAfterExport]({%slug grid-export-events%}#onafterexport) event of the Grid. It fires before the actual file is provided to the user. The `Stream` field of its event argument contains the output of the export as a `MemoryStream`.
```

> You can copy the bytes from this stream and import them in a `RadSpreadProcessing` [workbook](https://docs.telerik.com/devtools/document-processing/libraries/radspreadprocessing/working-with-workbooks/working-with-workbooks-what-is-workbook) to access and modify the file cells as needed.

Reviewer: ```suggestion
You can copy the bytes from this stream and import them in a [`RadSpreadProcessing` workbook](https://docs.telerik.com/devtools/document-processing/libraries/radspreadprocessing/working-with-workbooks/working-with-workbooks-what-is-workbook) to access and modify the cells.
```

[`RadSpreadProcessing`](https://docs.telerik.com/devtools/document-processing/libraries/radspreadprocessing/overview) is a powerful library that allows you to create spreadsheets from scratch, modify existing documents or convert between the most common spreadsheet formats. In this case, we will focus on the file modification.

The example below targets Excel file export and customization. Same approach can be applied for [CSV files](https://docs.telerik.com/devtools/document-processing/libraries/radspreadprocessing/formats-and-conversion/csv/csvformatprovider).

To customize the cell format of the exported file before it reaches the client, do the following:

> * Install `Telerik.Documents.Spreadsheet.FormatProviders.Xls` package for the `workbook` import.

Reviewer: Consider an ordered list here.

Author: Done

    <GridToolBar>
        <GridCommandButton Command="ExcelExport" Icon="file-excel">Export to Excel</GridCommandButton>
        <label class="k-checkbox-label"><TelerikCheckBox @bind-Value="@ExportAllPages" />Export All Pages</label>
    </GridToolBar>

    <GridExport>
        <GridExcelExport FileName="telerik-grid-export"
                         AllPages="@ExportAllPages"
                         OnAfterExport="@OnExcelAfterExport" />
    </GridExport>

    <GridColumns>
        <GridColumn Field="@nameof(SampleData.ProductId)" Title="ID" Width="100px" />
        <GridColumn Field="@nameof(SampleData.ProductName)" Title="Product Name" Width="300px" />
        <GridColumn Field="@nameof(SampleData.UnitsInStock)" Title="In stock" Width="100px" />
        <GridColumn Field="@nameof(SampleData.Price)" Title="Unit Price" Width="200px" />
        <GridColumn Field="@nameof(SampleData.Discontinued)" Title="Discontinued" Width="100px" />
        <GridColumn Field="@nameof(SampleData.FirstReleaseDate)" Title="Release Date" Width="300px" />
    </GridColumns>
</TelerikGrid>

@code {
    private async Task OnExcelAfterExport(GridAfterExcelExportEventArgs args)
    {

>         //and and copy the bytes to a new MemoryStream

Reviewer: Maybe something is missing here, in place of the duplicate "and"

Author: FIxed

        var bytes = args.Stream.ToArray();

        var excelStream = new MemoryStream(bytes);

        //create a format provider instance to call the import
        XlsxFormatProvider formatProvider = new XlsxFormatProvider();

        //import the stream to a workbook
        Workbook workbook = formatProvider.Import(excelStream);

>         //create cell selection

Reviewer: You can explain the code with a slightly different wording, for better understanding
```suggestion
        //select a range of cells
```
